ncm-spma/src/main/perl/spma/yumng.pm

```diff
@@ -0,0 +1,810 @@
+        }
+        my $vra = $pkgs->{$name};
+        while ( my ( $vers, $a ) = each(%$vra) ) {
+            my $arches = $a->{arch};
+            if ( exists( $a->{repository} ) ) {
+                foreach my $arch (@$arches) {
+                    if ( $vers ne '_' ) {
+                        push ( @pkl_v, ( unescape $name) . ';' . ( unescape $vers) . '.' . $arch );
+                    } else {
+                        if ( $arch eq '_' ) {
+                            push ( @pkl, unescape $name . ';' );
+                        } else {
+                            push ( @pkl_a, unescape $name . ';' . $arch );
+                        }
+                    }
+                }
+            } else {
+                foreach my $arch ( keys %$arches ) {
+                    if ( $vers ne '_' ) {
+                        push ( @pkl_v, ( unescape $name) . ';' . ( unescape $vers) . '.' . $arch );
+                    } else {
+                        if ( $arch eq '_' ) {
+                            push ( @pkl, unescape $name . ';' );
+                        } else {
+                            push ( @pkl_a, unescape $name . ';' . $arch );
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    if ( !$found_spma ) {
+        $self->error('Package ncm-spma is not present among required packages.');
+        return 0;
+    }
+
+    my $wanted_pkgs_uv = Set::Scalar->new(@pkl);
+    my $wanted_pkgs_v  = Set::Scalar->new(@pkl_v);
+    my $wanted_pkgs_a  = Set::Scalar->new(@pkl_a);
+    while ( defined( my $p = $wanted_pkgs_uv->each ) ) {
+        while ( defined( my $p2 = $wanted_pkgs_v->each ) ) {
+            if ( index( $p2, $p ) == 0 ) {
+                my $pkg = $p;
+                chop($pkg);
+                my $pkg_locked = $p2;
+                $pkg_locked =~ tr/;/-/;
+                $self->info("preferring version-locked $pkg_locked over unlocked $pkg");
+                $wanted_pkgs_uv->delete($p);
+            }
+        }
+    }
+    while ( defined( my $p = $wanted_pkgs_uv->each ) ) {
+        chop($p);
+        $wanted_pkgs->insert($p);
+    }
+    while ( defined( my $p = $wanted_pkgs_v->each ) ) {
+        $p =~ s/;/-/;
+        $wanted_pkgs_locked->insert($p);
+    }
+    while ( defined( my $p = $wanted_pkgs_a->each ) ) {
+        $p =~ s/;/./;
+        $wanted_pkgs->insert($p);
+    }
+
+    if ( !$NoAction ) {
+        # Import GPG keys
+        foreach my $file ( glob "/etc/pki/rpm-gpg/RPM-GPG-KEY*" ) {
+            my $gpg = $self->execute_yum_command( ["rpm -v --import $file"], "importing GPG key $file" );
+            return 0 if !defined($gpg);
+        }
+    }
+
+    # Get list of packages installed on system before any package modifications.
+    my $preinstalled_rpms = $self->execute_yum_command( [RPM_QUERY_INSTALLED], "getting list of installed packages", 1 );
+    return 0 if !defined($preinstalled_rpms);
+    $preinstalled_rpms =~ s/\(none\)/0/g;
```

Comment:
Sometimes the result of YUM transaction is not exactly defined beforehand. E.g yum downgrade with list of packages. The invocation is to get list of packages installed before and after yum transaction to get info what actually changed (and log it). It is useful to also check yum misbehaviour which is obvious from no package changes for example.